### PR TITLE
[Doc] [runtime env] Add runtime env architecture doc

### PR DESCRIPTION
**Pull request by architkulkarni**
Signed-off-by: Archit Kulkarni <architkulkarni@users.noreply.github.com>

## Why are these changes needed?
Adds `ARCHITECTURE.md` to the `runtime_env` directory, intended as a reference for new developers.  For now it's not self-contained, since there's useful information in existing docs and design docs (which are linked), but we can consider making it completely self-contained.

## Checks

- [ ] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by architkulkarni** [comment]
@edoakes @iycheng @rkooo567 Writing this instead of an internal knowledge transfer doc or talk; hopefully it's more useful this way as a public reference.  It's a rough draft right now and a bit of a brain dump, I'd appreciate your input on organization or things that are missing.

cc @gvspraveen

File: python/ray/runtime_env/ARCHITECTURE.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+Worker process are started by the Raylet in python/ray/_private/services.py. In this command, python/ray/_private/workers/setup_worker.py is run, which deserializes the `RuntimeEnvContext` object and calls its `exec_worker` method.  The `exec_worker` method sets the appropriate environment variables, modifies the worker process startup command (e.g. prepending `conda activate some_env`) and then calls `execvp` to start the worker process (python/ray/_private/worker.py).
+
+## Caching and garbage collection
+
+For design details on caching and garbage collection, see the [design doc](https://docs.google.com/document/d/1x1JAHg7c0ewcOYwhhclbuW0B0UC7l92WFkF4Su0T-dk/edit#heading=h.j4mqiaz83o96).
+
+The implementation is as follows:
+
+### GCS internal KV garbage collection
```
Comment (edoakes) [inline python/ray/runtime_env/ARCHITECTURE.md:50]: suggest to add explanation of the "temporary reference"

File: python/ray/runtime_env/ARCHITECTURE.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+References are tracked per package (URI).  The reference counting is managed in src/ray/common/runtime_env_manager.cc.  References to these files are incremented when a driver is started that uses that URI, and decremented when a driver exits.  Similarly, references are incremented when a detached actor is started that uses that URI, and decremented when that actor exits.
+
+When a reference count reaches 0, the file is deleted.
+
+### Local node garbage collection
+
+This section deals with files that are stored on disk on all nodes, such as installed `pip` packages or `working_dir` files downloaded from the GCS or from a remote URI.
+
+References for these files are tracked by the runtime env agent process on each node.  Each agent keeps its own reference table in memory.  References are incremented when a runtime env is created, and decremented when the runtime env is deleted.  The files are not deleted until the reference count reaches 0 and the cache size exceeds the maximum cache size.
```
Comment (edoakes) [inline python/ray/runtime_env/ARCHITECTURE.md:62]: max cache size defaults to XXX and can be set using YYY

File: python/ray/runtime_env/ARCHITECTURE.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+This document describes the architecture of Ray's runtime environment feature.
+
+## Overview
+
+For a high-level overview of runtime environments, see the blog post [here](https://www.anyscale.com/blog/handling-files-and-packages-on-your-cluster-with-ray-runtime-environments).  The blog post also contains a simple example of how to use runtime environments.
+
+For a more detailed explanation of how to use runtime environments, see the [documentation](https://docs.ray.io/en/latest/ray-core/handling-dependencies.html).
+
+## Architecture
```
Comment (edoakes) [inline python/ray/runtime_env/ARCHITECTURE.md:11]: suggest to add info about the runtime_env agent fate sharing with the raylet (and why that's important)

**Review by edoakes — COMMENTED** [review]
A figure describing the "life of a runtime_env" would be very helpful

File: python/ray/runtime_env/ARCHITECTURE.md (hunk 1 above)
Comment (architkulkarni) [reply python/ray/runtime_env/ARCHITECTURE.md:50]: Yup, added

File: python/ray/runtime_env/ARCHITECTURE.md (hunk 2 above)
Comment (architkulkarni) [reply python/ray/runtime_env/ARCHITECTURE.md:68]: Done

File: python/ray/runtime_env/ARCHITECTURE.md (hunk 3 above)
Comment (architkulkarni) [reply python/ray/runtime_env/ARCHITECTURE.md:11]: Added

**Review by edoakes — APPROVED** [review]
Nice!

**Comment by edoakes** [comment]
@architkulkarni I suggest you give this to someone who knows nothing about runtime_env and have them read it and ask questions... that'll be better feedback than from someone who already understands it.

**Comment by edoakes** [comment]
(maybe after merging)

**Comment by architkulkarni** [comment]
> @architkulkarni I suggest you give this to someone who knows nothing about runtime_env and have them read it and ask questions... that'll be better feedback than from someone who already understands it.

Good idea -- maybe @angelinalg would it make sense for you to review this?

**Comment by angelinalg** [comment]
I'm struggling a bit to understand what a Raylet is. From the blog, it's "process that runs on each node in a Ray cluster." Is it a process that runs only when you have runtime environments? If it's a process that runs every time you start a cluster, what does the process do? If this is obvious to our audience, then please ignore.

**Comment by architkulkarni** [comment]
@angelinalg thanks for the feedback, it makes sense!  I think Ray developers will be familiar with the Raylet, but it doesn't hurt to add a reference to the [Ray whitepaper](https://docs.google.com/document/d/1tBw9A4j62ruI5omIJbMxly-la5w4q_TjyJgJL_jN2fI/preview) so I'll do that.  I'll also add more details about the plugins, since that will not necessarily be familiar to new developers of runtime_env.

**Comment by architkulkarni** [comment]
Workflow tests flaky on master and unrelated (this PR just adds a `.md` file)

**Pull request merged**